src/adapters/SwivelAdapter.sol

```diff
@@ -0,0 +1,158 @@
+import {IERC20} from 'src/interfaces/IERC20.sol';
+import {IYield} from 'src/interfaces/IYield.sol';
+import {IMarketPlace} from 'src/interfaces/IMarketPlace.sol';
+
+import {Swivel} from 'src/lib/Swivel.sol';
+import {Exception} from 'src/errors/Exception.sol';
+
+// NOTE: The storage of the adapter _must_ exactly match the storage layout of the Lender
+contract SwivelAdapter is IAdapter {
+    /// @notice minimum wait before the admin may withdraw funds or change the fee rate
+    uint256 public constant HOLD = 3 days;
+
+    /// @notice address that is allowed to set and withdraw fees, disable principals, etc. It is commonly used in the authorized modifier.
+    address public admin;
+    /// @notice address of the MarketPlace contract, used to access the markets mapping
+    address public marketPlace;
+    /// @notice mapping that determines if a principal has been paused by the admin
+    mapping(uint8 => bool) public paused;
+    /// @notice flag that allows admin to stop all lending and minting across the entire protocol
+    bool public halted;
+
+    /// @notice protocol specific addresses that adapters reference when executing lends
+    /// @dev these addresses are references by an implied enum; adapters hardcode the index for their protocol
+    address[] public protocolAddressess;
+
+    /// @notice a mapping that tracks the amount of unswapped premium by market. This underlying is later transferred to the Redeemer during Swivel's redeem call
+    mapping(address => mapping(uint256 => uint256)) public premiums;
+
+    /// @notice this value determines the amount of fees paid on loans
+    uint256 public feenominator;
+    /// @notice represents a point in time where the feenominator may change
+    uint256 public feeChange;
+    /// @notice represents a minimum that the feenominator must exceed
+    uint256 public constant MIN_FEENOMINATOR = 500;
+
+    /// @notice maps underlying tokens to the amount of fees accumulated for that token
+    mapping(address => uint256) public fees;
+    /// @notice maps a token address to a point in time, a hold, after which a withdrawal can be made
+    mapping(address => uint256) public withdrawals;
+
+    // Reantrancy protection
+    uint256 private constant _NOT_ENTERED = 1;
+    uint256 private constant _ENTERED = 2;
+    uint256 private _status = _NOT_ENTERED;
+
+    // Rate limiting protection
+    /// @notice maximum amount of value that can flow through a protocol in a day (in USD)
+    uint256 public maximumValue = 250_000e27;
+    /// @notice maps protocols to how much value, in USD, has flowed through each protocol
+    mapping(uint8 => uint256) public protocolFlow;
+    /// @notice timestamp from which values flowing through protocol has begun
+    mapping(uint8 => uint256) public periodStart;
+    /// @notice estimated price of ether, set by the admin
+    uint256 public etherPrice = 2_500;
+
+    function approve(address[] calldata a) external {}
```

Comment:
I assume youre saving for a later PR?

Reply:
Yup, just wanted to iron out the lend, but a good call out.